doc/erofs.md

```diff
@@ -0,0 +1,431 @@
+# erofs: the missing manual
+
+## Introduction
+
+This is an attempt to document the format of erofs (or at least the subsets of
+it that we use in composefs).
+
+It probably makes sense to have `erofs_fs.h` open when reading this.
+
+## Overall concepts
+
+All integers (including all offsets) are stored in little-endian byte order.
+
+The file layout is fairly free-form.  You can freely mix inodes, data blocks,
+and shared xattr entries.  inodes are 64-bit values based on file offsets
+rather than integer indexes into a fixed table, so they can be anywhere at all.
+xattrs are 32-bit values based on offsets, so they're a bit more limited (but
+not in filesystems of reasonable size).
+
+## The first 1024 bytes (pre-superblock)
+
+The first 1024 bytes of an erofs have no particular meaning.  You can put
+anything you want there, like partition tables or boot sectors or anything
+else.  composefs puts its own header inside of this area, at the start.
+
+## The superblock (at 1024 bytes, 128 bytes long)
+
+The superblock is defined by `struct erofs_super_block`.
+
+Here's some notes about some of the fields.  Anything not mentioned is left as
+0 by us. There's some pretty wild features in here, but we don't use them all
+(and I don't understand them, either) so they're not all documented.
+
+*   `magic`: set that to `EROFS_SUPER_MAGIC_V1` (`0xE0F5E1E2`)
+*   `checksum`: only meaningful of the `SB_CHKSUM` feature is enabled.  This is
+    a crc32c over a block-sized-chunk of data starting from the superblock,
+    with this field set to 0.  That's pretty weird.  Maybe don't use this.
+*   `feature_compat`: a flags field.  The filesystem will still mount even if
+    the kernel doesn't know about any features which might be present.  The
+    flags:
+    -   `SB_CHKSUM` (`0x0001`): set if the checksum field in the superblock is
+        populated.  Otherwise, the checksum is ignored.
+    -   `MTIME` (`0x0002`): at first, erofs named the timestamp fields `ctime`
+        instead of `mtime`.  That got changed a long time ago, and this flag
+        got added to indicate filesystems that were created with the new
+        semantics. This flag has absolutely zero impact at run time: the kernel
+        ignores it.
+    -   `XATTR_FILTER` (`0x0004`): set if the xattr bloom filter should be
+        used.  Read about this in the inode section.
+*   `blkszbits`: log2 of the block size.  Better set this to 12 (4096).
```

Comment:
yes, also it may be helpful to indicate that it's useful to make the image mountable on any page sizes.

Reply:
Can you please clarify what you meant by this?  Should the block size be larger or smaller than the page size in order to be able to mount it?

Reply:
on the modern platforms, the smallist page size is 4KiB. EROFS has supported < page size blocks for all platforms now. So leave it as 4KiB will keep images mountable all the time..